src/replication.c

```diff
@@ -2308,6 +2474,441 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
+        server.repl_rdb_transfer_s = NULL;
+    }
+    if (server.repl_transfer_tmpfile) {
+        zfree(server.repl_transfer_tmpfile);
+        server.repl_transfer_tmpfile = NULL;
+    }
+    if (server.repl_transfer_fd != -1) {
+        close(server.repl_transfer_fd);
+        server.repl_transfer_fd = -1;
+    }
+    server.repl_rdb_conn_state = REPL_RDB_CONN_STATE_NONE;
+    server.repl_provisional_primary.read_reploff = 0;
+    server.repl_provisional_primary.reploff = 0;
+    server.repl_provisional_primary.conn = NULL;
+    server.repl_provisional_primary.dbid = -1;
+    server.rdb_client_id = -1;
+    freePendingReplDataBuf();
+    return;
+}
+
+/* Replication: Primary side.
+ * Send current replication offset to replica. Use the following structure:
+ * $ENDOFF:<repl-offset> <primary-repl-id> <current-db-id> <client-id> */
+int sendCurrentOffsetToReplica(client *replica) {
+    char buf[128];
+    int buflen;
+    buflen = snprintf(buf, sizeof(buf), "$ENDOFF:%lld %s %d %llu\r\n", server.primary_repl_offset, server.replid,
+                      server.db->id, (long long unsigned int)replica->id);
+    serverLog(LL_NOTICE, "Sending to replica %s RDB end offset %lld and client-id %llu",
+              replicationGetReplicaName(replica), server.primary_repl_offset, (long long unsigned int)replica->id);
+    if (connSyncWrite(replica->conn, buf, buflen, server.repl_syncio_timeout * 1000) != buflen) {
+        freeClientAsync(replica);
+        return C_ERR;
+    }
+    return C_OK;
+}
+
+/* Replication: Replica side.
+ * This connection handler is used to initialize the RDB connection (dual-conn-sync-enabled sync).
+ * Once a replica with repl rdb-connection enabled, denied from PSYNC with its primary,
+ * fullSyncWithPrimary begins its role. The connection handler prepare server.repl_rdb_transfer_s
+ * for a rdb stream, and server.repl_transfer_s for increamental replication data stream. */
+void fullSyncWithPrimary(connection *conn) {
+    char *err = NULL;
+    serverAssert(conn == server.repl_rdb_transfer_s);
+    /* If this event fired after the user turned the instance into a primary
+     * with SLAVEOF NO ONE we must just return ASAP. */
+    if (server.repl_state == REPL_STATE_NONE) {
+        goto error;
+    }
+    /* Check for errors in the socket: after a non blocking connect() we
+     * may find that the socket is in error state. */
+    if (connGetState(conn) != CONN_STATE_CONNECTED) {
+        serverLog(LL_WARNING, "Error condition on socket for RDB-CONNECTION-SYNC: %s", connGetLastError(conn));
+        goto error;
+    }
+    /* Send replica capabilities */
+    if (server.repl_rdb_conn_state == REPL_RDB_CONN_SEND_HANDSHAKE) {
+        serverLog(LL_DEBUG, "Received first reply from primary using rdb connection.");
+        /* AUTH with the primary if required. */
+        if (server.primary_auth) {
+            char *args[3] = {"AUTH", NULL, NULL};
+            size_t lens[3] = {4, 0, 0};
```

Comment:
nit - can save one letter :)
```suggestion
            char *args[] = {"AUTH", NULL, NULL};
            size_t lens[] = {4, 0, 0};
```

Reply:
👍